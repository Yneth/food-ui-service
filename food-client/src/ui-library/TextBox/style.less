@import "../styles/variables";

@border-color: lighten(@brand-primary, 30%);

.text-box__input {
  padding: .5em .8em;

  border: 1px solid @gray-light;

  border-radius: 4px;
  transition: border-color .3s, box-shadow .3s;

  &:focus {
    border-color: @border-color;
    outline: none;
    box-shadow: 0 0 8px @border-color;
  }
}

.text-box__label {
  display: block;
  margin: .5em 0;

  font-weight: 700;
}

.text-box__error {
  display: block;
  font-size: 0.9em;
  color: @brand-danger;
}