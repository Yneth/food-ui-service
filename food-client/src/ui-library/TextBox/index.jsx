import React from 'react';
import PropTypes from 'prop-types';
import getId from 'services/getId';
import './style.less';

class TextBox extends React.Component {

    constructor(props) {
        super(props);

        this.id = getId();

        this.state = {
            value: props.defaultValue,
        };

        this.handleChange = this.handleChange.bind(this);
    }

    handleChange(e) {
        const { onChange } = this.props;
        const { value } = e.target;

        this.setState({ value });

        if (onChange) {
            onChange(value);
        }
    }

    renderLabel() {
        const { label } = this.props;

        return label ?
            <label className="text-box__label" htmlFor={this.id}>{ label }</label>
            :
            null;
    }

    renderError() {
        const { error } = this.props;

        return error ? <span className="text-box__error">{ error }</span> : null;
    }

    renderInput(props) {
        const { type } = this.props;

        if (type === 'area') {
            return <textarea {...props} />;
        }

        return <input type={type} {...props} />;
    }

    render() {
        const { inputRef, placeholder } = this.props;
        const { value } = this.state;

        const props = {
            id: this.id,
            ref: inputRef,
            className: 'text-box__input',
            onChange: this.handleChange,
            value,
            placeholder,
        };

        return (
            <div>
                { this.renderLabel() }
                { this.renderInput(props) }
                { this.renderError() }
            </div>
        );
    }
}

TextBox.defaultProps = {
    type: 'text',
    defaultValue: '',
};

TextBox.propTypes = {
    onChange: PropTypes.func,
    inputRef: PropTypes.func,
    type: PropTypes.oneOf(['text', 'password', 'area']),
    defaultValue: PropTypes.string,
    label: PropTypes.string,
    error: PropTypes.string,
    placeholder: PropTypes.string,
};

export default TextBox;
