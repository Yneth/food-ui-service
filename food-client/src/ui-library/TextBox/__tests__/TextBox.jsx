import React from 'react';
import renderer from 'react-test-renderer';
import { mount } from 'enzyme';
import TextBox from '../';

describe('TextBox component', () => {
    it('should render input', () => {
        const tree = renderer.create(<TextBox />).toJSON();
        expect(tree).toMatchSnapshot();
    });

    it('should render textarea', () => {
        const tree = renderer.create(<TextBox type="area" />).toJSON();
        expect(tree).toMatchSnapshot();
    });

    it('should render input with customization', () => {
        const tree = renderer.create(
            <TextBox
                defaultValue="val"
                label="INPUT:"
                error="NOT VALID"
                className="custom class"
            />,
        ).toJSON();
        expect(tree).toMatchSnapshot();
    });

    it('should call handler on change event', () => {
        const handler = jest.fn();
        const wrapper = mount(<TextBox onChange={handler}>Hello</TextBox>);
        wrapper.find('input').simulate('change', { target: { value: 'hello' } });

        expect(handler).toHaveBeenCalledWith('hello');
    });
});

