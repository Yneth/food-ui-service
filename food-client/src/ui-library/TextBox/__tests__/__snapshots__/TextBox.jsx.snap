// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TextBox component should render input 1`] = `
<div>
  <input
    className="text-box__input"
    id={0}
    onChange={[Function]}
    placeholder={undefined}
    type="text"
    value=""
  />
</div>
`;

exports[`TextBox component should render input with customization 1`] = `
<div>
  <label
    className="text-box__label"
    htmlFor={2}
  >
    INPUT:
  </label>
  <input
    className="text-box__input custom class"
    id={2}
    onChange={[Function]}
    placeholder={undefined}
    type="text"
    value="val"
  />
  <span
    className="text-box__error"
  >
    NOT VALID
  </span>
</div>
`;

exports[`TextBox component should render textarea 1`] = `
<div>
  <textarea
    className="text-box__input"
    id={1}
    onChange={[Function]}
    placeholder={undefined}
    value=""
  />
</div>
`;
